balancer/xds/xds.go

```diff
@@ -220,41 +221,52 @@ func (x *xdsBalancer) run() {
 
 func (x *xdsBalancer) handleGRPCUpdate(update interface{}) {
 	switch u := update.(type) {
-	case *addressUpdate:
-		if x.fallbackLB != nil {
-			x.fallbackLB.HandleResolvedAddrs(u.addrs, u.err)
-		}
-		x.fallbackInitData = u
 	case *subConnStateUpdate:
 		if x.xdsLB != nil {
 			x.xdsLB.HandleSubConnStateChange(u.sc, u.state)
 		}
 		if x.fallbackLB != nil {
 			x.fallbackLB.HandleSubConnStateChange(u.sc, u.state)
 		}
-	case *xdsConfig:
-		if x.config == nil {
-			// The first time we get config, we just need to start the xdsClient.
-			x.startNewXDSClient(u)
-			x.config = u
-			return
-		}
-		// With a different BalancerName, we need to create a new xdsClient.
-		// If current or previous ChildPolicy is nil, then we also need to recreate a new xdsClient.
-		// This is because with nil ChildPolicy xdsClient will do CDS request, while non-nil won't.
-		if u.BalancerName != x.config.BalancerName || (u.ChildPolicy == nil) != (x.config.ChildPolicy == nil) {
-			x.startNewXDSClient(u)
+	case *resolverUpdate:
+		if u.addrUpdate != nil {
+			// addresses have been updated.
+			// in case of x.config == nil where it returns early, we set the fallbackInitData here.
+			x.fallbackInitData = u.addrUpdate
```

Comment:
Why do you still need `fallbackInitData`?

If fallback balancer is updated, the new addresses are also sent along with the update.
You shouldn't need to keep the old addresses for a new fallback balancer.

Reply:
Actually, you need this field when switching to fallback.
But this update here shouldn't be necessary. It also causes double address update for a new fallback balancer.

Reply:
It's here because the early return. Let me think a bit more if we can just move this line inside the early return condition.